lib/tcp.c

```diff
@@ -0,0 +1,189 @@
+    channel_state->expected_recv_msg_seq_index = INITIAL_RECV_MSG_SEQ_INDEX;
+    return channel_state;
+}
+
+void channel_state_free(ChannelState* channel_state)
+{
+    queue_free(channel_state->send_queue);
+    free(channel_state);
+}
+
+void channel_state_free_void(void* channel_state)
+{
+    channel_state_free(channel_state);
+}
+
+void tcp_init(const char** peers, const size_t num_peers)
+{
+    CHANNEL_STATES = dictionary_init(channel_state_free_void);
+    PEERS = malloc(num_peers * sizeof(char*));
+    for (size_t i = 0; i < num_peers; i++)
+    {
+        dictionary_set(CHANNEL_STATES, peers[i], channel_state_init());
+        PEERS[i] = strdup(peers[i]);
+        NUM_PEERS++;
+    }
+}
+
+void tcp_free(void)
+{
+    if (CHANNEL_STATES == NULL)
+    {
+        fprintf(stderr, "Error: tcp_free: tcp_init not called\n");
+        exit(EXIT_FAILURE);
+    }
+    for (size_t i = 0; i < NUM_PEERS; i++)
+    {
+        ChannelState* channel_state = dictionary_get(CHANNEL_STATES, PEERS[i]);
+        queue_free(channel_state->send_queue);
+        free(PEERS[i]);
+    }
+    free(PEERS);
+    dictionary_free(CHANNEL_STATES);
+}
+
+// Borrows `msg`.
+static json_object* generate_ACK_msg(json_object* msg)
+{
+    json_object* ack_msg = generic_reply(msg);
+    json_object* ack_body = json_object_object_get(ack_msg, "body");
+    json_object_object_add(ack_body, "type", json_object_new_string("ACK"));
+    json_object* body = json_object_object_get(msg, "body");
+    json_object* seq_msg_index = json_object_object_get(body, "seq_msg_index");
+    json_object_get(seq_msg_index);
+    json_object_object_add(ack_body, "seq_msg_index", seq_msg_index);
+    return ack_msg;
+}
+
+void msg_send_pusher(json_object* msg)
+{
+    fprintf(stderr, "(msg_send_pusher) start\n");
+    const char* peer =
+        json_object_get_string(json_object_object_get(msg, "dest"));
+    Message* m = malloc(sizeof(Message));
+    fprintf(stderr, "(push) before get. peer: %s\n", peer);
+    ChannelState* channel_state = dictionary_get(CHANNEL_STATES, peer);
+    fprintf(stderr, "(push) after get\n");
+    m->msg_seq_index = channel_state->next_send_msg_seq_index++;
+    json_object* body = json_object_object_get(msg, "body");
+    json_object_object_add(body, "seq_msg_index",
+                           json_object_new_uint64(m->msg_seq_index));
+    m->msg = msg;
+    queue_enqueue(channel_state->send_queue, m);
+}
+
+json_object* msg_recv_listener()
+{
+    while (true)
+    {
+        for (size_t i = 0; i < NUM_PEERS; i++)
+        {
+            ChannelState* channel_state =
+                dictionary_get(CHANNEL_STATES, PEERS[i]);
+            if (!queue_is_empty(channel_state->send_queue))
+            {
+                Message* m = queue_peek(channel_state->send_queue);
+                json_object_get(m->msg);
+                fprintf(stderr, "(send) msg: %s\n",
+                        json_object_to_json_string(m->msg));
+                msg_send(m->msg);
```

Comment:
@khollbach , here's what I think is going on re: duplicate self-messages (with an example at the bottom):

1. Send a queued "ping" (it was placed into the send queue by the application code)
2. Receive the "ping" on L136
3. Process it as a data message, so send an "ACK" message, return the data message to application code (prompting "pong" to be enqueued to the send queue).
4. `msg_recv_listener` is called again, "ping" is sent *again* before the next message (ACK from Step 3) is processed
5. Process the ACK from Step 3 accordingly (importantly, dequeuing the "ping" from the send queue) and stay in `msg_recv_listener` `while` loop...
6. ...which now sends the enqueued "pong" to itself and receives the ACK for the second "ping" (from Step 4), ignoring it and putting us back at the top of the `msg_recv_listener` while loop...
7. ...which sends "pong" *again*, but what get's processed on L136 is the *first* "pong."
8. Process it as a data message, so send an "ACK" message, return the data message to application code (prompting the client response to be sent).
9. ...but we still have things in STDIN! Next up is the *second* "pong", but before that, we send *another* "pong" before processing the *second* "pong" (which generates another "ACK" message).

...

Here's this same example seen in The Actual Data:

```
c12 → n1 {:echo "Please echo 17", :type "echo", :msg_id 7}
n1 → n1 {:in_reply_to 7, :type "ping", :original_msg {"id" 274, "src" "c12", "dest" "n1", "body" {"echo" "Please echo 17", "type" "echo", "msg_id" 7}}, :seq_msg_index 2, :msg_id 45}
n1 → n1 {:in_reply_to 45, :type "ACK", :seq_msg_index 2, :msg_id 46}
n1 → n1 {:in_reply_to 7, :type "ping", :original_msg {"id" 274, "src" "c12", "dest" "n1", "body" {"echo" "Please echo 17", "type" "echo", "msg_id" 7}}, :seq_msg_index 2, :msg_id 47}
n1 → n1 {:in_reply_to 45, :type "pong", :original_msg {"id" 274, "src" "c12", "dest" "n1", "body" {"echo" "Please echo 17", "type" "echo", "msg_id" 7}}, :seq_msg_index 3, :msg_id 48}
n1 → n1 {:in_reply_to 47, :type "ACK", :seq_msg_index 2, :msg_id 49}
n1 → n1 {:in_reply_to 45, :type "pong", :original_msg {"id" 274, "src" "c12", "dest" "n1", "body" {"echo" "Please echo 17", "type" "echo", "msg_id" 7}}, :seq_msg_index 3, :msg_id 50}
n1 → n1 {:in_reply_to 48, :type "ACK", :seq_msg_index 3, :msg_id 51}
n1 → n1 {:in_reply_to 45, :type "pong", :original_msg {"id" 274, "src" "c12", "dest" "n1", "body" {"echo" "Please echo 17", "type" "echo", "msg_id" 7}}, :seq_msg_index 3, :msg_id 53}
n1 → c12 {:in_reply_to 7, :type "echo_ok", :echo "Please echo 17", :msg_id 52}
n1 → n1 {:in_reply_to 45, :type "pong", :original_msg {"id" 274, "src" "c12", "dest" "n1", "body" {"echo" "Please echo 17", "type" "echo", "msg_id" 7}}, :seq_msg_index 3, :msg_id 54}
n1 → n1 {:in_reply_to 50, :type "ACK", :seq_msg_index 3, :msg_id 55}
n1 → n1 {:in_reply_to 45, :type "pong", :original_msg {"id" 274, "src" "c12", "dest" "n1", "body" {"echo" "Please echo 17", "type" "echo", "msg_id" 7}}, :seq_msg_index 3, :msg_id 56}
n1 → n1 {:in_reply_to 53, :type "ACK", :seq_msg_index 3, :msg_id 57}
n1 → n1 {:in_reply_to 54, :type "ACK", :seq_msg_index 3, :msg_id 58}
n1 → n1 {:in_reply_to 56, :type "ACK", :seq_msg_index 3, :msg_id 59}
```

Reply:
wow, excellent sleuthing! It's kind of surprising that only self-messages triggered this problem... thanks for getting to the bottom of it!

As far as how to fix it... I have some ideas (timeouts!) but the details are fuzzy. Let's discuss next time we meet!